### PR TITLE
Add send queue

### PR DESCRIPTION
**Pull request by tank157**
[Pivotal Card](https://www.pivotaltracker.com/story/show/118434359)

### CODE REVIEW

**Comment by tank157** [comment]
Assigning to @sduckett 

Note this is one of 3 PRs for the work, but I'm spreading them out.

**Comment by tank157** [comment]
@sduckett I'm happy to talk over this code in person too. Some of what it does is about following patterns from notification-works, and the general low amount of testing is about the very side-effecting nature of the work.

File: src/mail_works/pdf.clj [hunk 1]
```diff
@@ -0,0 +1,63 @@
+(ns mail-works.pdf
+  (:require [pdfboxing.merge :as pdf]
+            [clj-http.client :as http]
+            [clojure.string :as string]
+            [clojure.java.io :as io]
+            [com.climate.newrelic.trace :refer [defn-traced]])
+  (:import [java.nio.file.attribute FileAttribute]
+           [java.nio.file Files]))
+
+(defn base-filename
```
Comment (sduckett) [inline src/mail_works/pdf.clj]: Would something like http://raynes.github.io/fs/me.raynes.fs.html#var-base-name maybe fit nicely here?

Comment (tank157) [inline src/mail_works/pdf.clj]: Hmmmm, maybe. But that library seems concerned with filesystem files, and this function is starting from a URI. It _might_ still work, but seems a bit wrong. But there does seem to be a lot of useful things in that library we might want to use in the future.

File: src/mail_works/handlers/send.clj [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+(defn schedule
+  [schedule-fn message]
+  (log/debug "Scheduling message:" message)
+  (s/validate schedule-schema message)
+  (schedule-fn (->schedule-message message)))
+
+(defn schedule-or-send
+  [schedule-fn message]
+  (log/debug "Schedule-or-send message:" message)
+  (if (contains? message :send-after)
```
Comment (sduckett) [inline src/mail_works/handlers/send.clj]: If the message doesn't contain `:send-after`, can we safely assume that it's in the correct form to send? The only schema check comes when we `schedule` a message.

Comment (tank157) [inline src/mail_works/handlers/send.clj]: Ah, good catch, there should still be some validation. Let me take care of that.

File: src/mail_works/pdf.clj [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+  [[the-name] & body]
+  `(let [temp-dir# (-> (str (java.util.UUID/randomUUID))
+                       (Files/createTempDirectory (into-array FileAttribute []))
+                       .toFile)
+         ~the-name temp-dir#]
+     (try
+       (do ~@body)
+       (finally (delete-dir temp-dir#)))))
+
+(defn-traced merge-pdfs
```
Comment (sduckett) [inline src/mail_works/pdf.clj]: What happens if we forget to provide filehandles and give this strings instead?

Comment (tank157) [inline src/mail_works/pdf.clj]: Well, merge-pdfs is just a thin wrapper on the pdfboxing libraries version, which does expect string. So if you provide a string already, it'll call `str` on it which should be a no-op.

File: src/mail_works/sftp.clj [hunk 4]
```diff
@@ -0,0 +1,22 @@
+(ns mail-works.sftp
+  (:require [turbovote.resource-config :refer [config]]
+            [clj-ssh.ssh :as ssh]
+            [com.climate.newrelic.trace :refer [defn-traced]]))
+
+(defn-traced upload
+  "Uploads the file (should be a File object) to the
```
Comment (sduckett) [inline src/mail_works/sftp.clj]: "Uploads the file (should be a File object)..." what happens if it's not a `File` like it should be? `with-open` fails?

Comment (tank157) [inline src/mail_works/sftp.clj]: Yeah, I think so. I'm going to take another look at this code here in a few, I think I may not be using `with-open` quite properly.

File: src/mail_works/sftp.clj [hunk 5]
```diff
@@ -9,7 +10,7 @@
    naming the resulting file with filename and in the
    supplied directory."
   [server username password file filename directory]
-  (with-open [stream file]
+  (with-open [stream (io/input-stream file)]
```
Comment (tank157) [inline src/mail_works/sftp.clj]: @sduckett I tweaked things to use `with-open` more idomatically (and so what's actually getting passed in matches the docstring).

**Comment by tank157** [comment]
@sduckett Satisfied with the most recent changes?

File: src/mail_works/sftp.clj (hunk 5 above)
Comment (sduckett) [inline src/mail_works/sftp.clj]: 👍

**Comment by sduckett** [comment]
@tank157 LGTM.

**Pull request merged**